### PR TITLE
Add SkipCA and SkipCN check requirement to WinRM/OMI HTTPS connection

### PR DESCRIPTION
**Pull request by PaulHigin**
## PR Summary

OMI does not support server certificate validation, so this change is to add a requirement that -SkipCACheck and -SkipCNCheck options are specified before a connection is allowed.  This way users will be aware that no server certificate validation is performed.

This **only** applies to PSCore6 remoting clients on non-Windows platforms, when creating a remote session to a Windows machine via WinRM\OMI.

This is a breaking change because these skip check options are now required.  If the skip check options are not included then the connection will be denied.

Examples:
```powershell
# On a Linux or MacOS machine
New-PSSession -Cn <WindowsMachine> -Credential $cred -Authentication Basic -UseSSL
New-PSSession : HTTPS on Unix does not support CA or CN checks. Use the PSSessionOption -SkipCACheck and -SkipCNCheck if you are certain of the server you are connecting to.
At line:1 char:1
+ New-PSSession -cn paulhig-2 -Credential $cred -Authentication Basic - ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ResourceUnavailable: (:) [New-PSSession], PSRemotingTransportException
+ FullyQualifiedErrorId : System.Management.Automation.Remoting.PSRemotingDataStructureException,Microsoft.PowerShell.Commands.NewPSSessionCommand

$so = New-PSSessionOption -SkipCACheck -SkipCNCheck
$session = New-PSSession -Cn <WindowsMachine> -Credential $cred -Authentication Basic -UseSSL -SessionOption $so

Enter-PSSession -Cn <WindowsMachine> -Credential $cred -Authentication Basic -UseSSL -SessionOption $so

```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/3299
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by adityapatwardhan** [comment]
@PaulHigin Please resolve merge conflict.

Also, enable tests here for non-windows: https://github.com/PowerShell/PowerShell/blob/master/test/powershell/engine/Remoting/SessionOption.Tests.ps1

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]
Please enable SessionOption tests.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by PaulHigin** [comment]
SessionOption.Tests.ps1 contains tests for WSMan SessionOption object, which we do not expose for non-Windows platforms.  But I'll add tests for New-PSSessionOption.

**Comment by iSazonov** [comment]
PowerShell Committee for the breaking change?

**Comment by TravisEz13** [comment]
Adding the committee to review the breaking change

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        It "<Name>" -TestCases $testCases {
+            param ($scriptBlock, $expectedErrorCode)
+
+            try {
+                & $scriptBlock
+                throw "No Exception!"
+            }
+            catch {
+                $_.Exception.ErrorCode | Should Be $expectedErrorCode
```
Comment (adityapatwardhan) [inline test/powershell/engine/Remoting/PSSession.Tests.ps1]: Please use Pester v4 syntax. `Be` -> `-Be`

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }
+    }
+}
+finally
+{
+    $PSDefaultParameterValues.remove("it:skip")
```
Comment (adityapatwardhan) [inline test/powershell/engine/Remoting/PSSession.Tests.ps1]: Please have a look at: https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md#skipping-tests-in-bulk for using PSDefaultParamterValues

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            },
+            @{
+                Name = 'Verifies expected error when SkipCACheck option is missing'
+                ScriptBlock = { New-PSSession -cn localhost -Credential $cred -Authentication Basic -UseSSl -SessionOption $soSkipCN }
+                ExpectedErrorCode = 825
+            },
+            @{
+                Name = 'Verifies expected error when SkipCNCheck option is missing'
+                ScriptBlock = { New-PSSession -cn localhost -Credential $cred -Authentication Basic -UseSSl -SessionOption $soSkipCA }
+                ExpectedErrorCode = 825
```
Comment (adityapatwardhan) [inline test/powershell/engine/Remoting/PSSession.Tests.ps1]: Use the FullyQualifiedErrorId for checking the expected error.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+            { $null = $cmdInfo.ResolveParameter("SkipCACheck") } | Should -Not -Throw -Because "SkipCACheck parameter should be available"
+            { $null = $cmdInfo.ResolveParameter("SkipCNCheck") } | Should -Not -Throw -Because "SkipCNCheck parameter should be available"
+        }
+    }
+
+    Describe "SkipCACheck and SkipCNCheck PSSession options are required for New-PSSession on non-Windows platforms" -Tag "CI" {
+
+        BeforeAll {
+            $cred = [pscredential]::new("BogusUser", (ConvertTo-SecureString -String "BogusPassword" -AsPlainText -Force))
```
[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: Fixed.

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                Name = 'Verifies expected error when SkipCNCheck option is missing'
+                ScriptBlock = { New-PSSession -cn localhost -Credential $cred -Authentication Basic -UseSSl -SessionOption $soSkipCA }
+                ExpectedErrorCode = 825
+            }
+        )
+
+        It "<Name>" -TestCases $testCases {
+            param ($scriptBlock, $expectedErrorCode)
+
+            try {
```
[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: I want to check for specific error code.  The ErrorId is too generic.

[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: Oh, I see what you mean.  Yes, I can check for both.

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 (hunk 1 above)
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: Whoops.  Fixed.

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 (hunk 3 above)
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: FulllyQualifiedErrorId is not specific enough.  I will check for both.

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 (hunk 2 above)
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: I am following patterns used in other tests.  If they are incorrect then they should be fixed.  It is my understanding that Before/After is not reliable and so try/finally is preferred when changing session state that affect all tests.

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 (hunk 3 above)
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: No, that won't work.  I need to check for specific error code, and will use existing pattern.

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 6]
```diff
@@ -4,66 +4,81 @@
...
-            }
-            catch {
-                $_.Exception.ErrorCode | Should Be $expectedErrorCode
-            }
+        try {
+            & $scriptBlock
+            throw "No Exception!"
+        }
+        catch {
+            $_.Exception.ErrorCode | Should -Be $expectedErrorCode
```
[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: I like this pattern, however the ErrorId is not being checked.  At least in my current version of Pester (4.4.2).

[the remark this answers is not in the document]
Comment (iSazonov) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: Sample from the repo https://github.com/PowerShell/PowerShell/blob/43e0394d51b7ee8c5fa84f5308e27cec4f942672/test/powershell/Language/Classes/Scripting.Classes.Attributes.Tests.ps1#L270-L274

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 (hunk 3 above)
Comment (TravisEz13) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: @adityapatwardhan Please update your review

File: src/System.Management.Automation/resources/RemotingErrorIdStrings.resx [hunk 7]
```diff
@@ -1670,4 +1670,7 @@ All WinRM sessions connected to PowerShell session configurations, such as Micro
   <data name="CannotGetHostInteropTypes" xml:space="preserve">
     <value>Host system does not have the correct version of Hyper-V schema.</value>
   </data>
+  <data name="UnixOnlyHttpsWithoutSkipCACheckNotSupported" xml:space="preserve">
+    <value>HTTPS on Unix does not currently support CA or CN checks. Use the PSSessionOption -SkipCACheck and -SkipCNCheck if you are certain of the server you are connecting to.</value>
```
Comment (TravisEz13) [inline src/System.Management.Automation/resources/RemotingErrorIdStrings.resx]:
```suggestion
    <value>HTTPS on Unix does not currently support CA or CN checks. Use the PSSessionOption -SkipCACheck and -SkipCNCheck if you are certain you trust the server you are connecting to and the network in between.</value>
```

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
One comment about the message

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Review by TravisEz13 — APPROVED** [review]

File: test/powershell/engine/Remoting/PSSession.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    AfterAll {
+        $global:PSDefaultParameterValues = $originalDefaultParameterValues
+    }
+
+    $testCases = @(
+        @{
+            Name = 'Verifies expected error when session options is missing'
+            ScriptBlock = { New-PSSession -cn localhost -Credential $cred -Authentication Basic -UseSSL }
```
[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/engine/Remoting/PSSession.Tests.ps1]: -cn is universally understood to be computer name.

**Comment by adityapatwardhan** [comment]
I thought `-cn` would flag a PSScriptAnalyzer error, but verified that it does not. So signing off.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@PowerShell/powershell-committee I did not realize this was marked for committee review. If this does not pass review, I will revert.

**Comment by TravisEz13** [comment]
@PowerShell/powershell-committee 	Also, consider if this should be back-ported to 6.1 and 6.0?

**Pull request merged**